lib/Dialect/FIRRTL/Transforms/ResolveTraces.cpp

```diff
@@ -126,30 +180,57 @@ struct ResolveTracesPass : public ResolveTracesBase<ResolveTracesPass> {
           })
           .Default([&](auto) { targetFieldID = 0; });
     }
+  }
+
+  /// Internal implementation that updates an Annotation to add a "target" field
+  /// based on the current location of the annotation in the circuit.  The value
+  /// of the "target" will be a local target if the Annotation is local and a
+  /// non-local target if the Annotation is non-local.
+  AnnoPathValue updateTargetImpl(Annotation &anno, FModuleLike &module,
+                                 FIRRTLBaseType type, hw::InnerRefAttr name,
+                                 AnnoTarget target) {
+    SmallString<64> newTarget("~");
+    newTarget.append(module->getParentOfType<CircuitOp>().getName());
+    newTarget.append("|");
 
-    anno.setMember("target", StringAttr::get(module->getContext(), newTarget));
+    SmallVector<InstanceOp> instances;
+
+    if (auto nla = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal")) {
+      hw::HierPathOp path = nlaTable->getNLA(nla.getAttr());
+      for (auto part : path.getNamepath().getValue().drop_back()) {
+        auto inst = cast<hw::InnerRefAttr>(part);
+        instances.push_back(dyn_cast<InstanceOp>(
+            istc->getInnerSymbolTable(nlaTable->getModule(inst.getModule()))
+                .lookupOp(inst.getName())));
+      }
+    }
 
-    LLVM_DEBUG(llvm::dbgs() << "    after:  " << anno.getDict() << "\n");
+    AnnoPathValue path(instances, target, anno.getFieldID());
 
-    return true;
+    return path;
   }
 
   /// Add a "target" field to a port Annotation that indicates the current
   /// location of the port in the circuit.
-  bool updatePortTarget(FModuleLike &module, Annotation &anno,
-                        unsigned portIdx) {
+  std::optional<AnnoPathValue>
+  updatePortTarget(FModuleLike &module, Annotation &anno, unsigned portIdx) {
     auto type = getBaseType(type_cast<FIRRTLType>(module.getPortType(portIdx)));
-    return updateTargetImpl(anno, module, type, module.getPortName(portIdx));
+    return updateTargetImpl(
+        anno, module, type,
+        hw::InnerRefAttr::get(module.getModuleNameAttr(),
+                              module.getPortSymbolAttr(portIdx).getSymName()),
+        PortAnnoTarget(module, portIdx));
   }
 
   /// Add a "target" field to an Annotation that indicates the current location
   /// of a component in the circuit.
-  bool updateTarget(FModuleLike &module, Operation *op, Annotation &anno) {
+  std::optional<AnnoPathValue> updateTarget(FModuleLike &module, Operation *op,
+                                            Annotation &anno) {
 
     // If this operation doesn't have a name, then do nothing.
-    StringAttr name = op->getAttrOfType<StringAttr>("name");
-    if (!name)
-      return false;
+    auto innerSym = op->getAttrOfType<hw::InnerSymAttr>("inner_sym");
```

Comment:
This should be redundant with the following access through `InnerSymbolOpInterface` - if it's /required/ (for this to work) that the target already has an inner symbol then just ask it via the interface.

Reply:
Yep, totally redundant.